\documentclass[handout,10pt]{beamer}

\usetheme{AnnArbor}
\usecolortheme{beaver}

\graphicspath{{include/}}

\setlength{\unitlength}{\textwidth}  % measure in textwidths
\usepackage[normalem]{ulem}

% to avoid counting all pages
\newcommand{\backupbegin}{
   \newcounter{finalframe}
   \setcounter{finalframe}{\value{framenumber}}
}
\newcommand{\backupend}{
   \setcounter{framenumber}{\value{finalframe}}
}


\setbeamertemplate{navigation symbols}{}
\setbeamertemplate{enumerate items}[default]
\setbeamertemplate{enumerate subitem}{\alph{enumii}.}
\setbeamertemplate{enumerate subsubitem}{\roman{enumiii}.}
\setkeys{Gin}{width=0.6\textwidth}

\newcommand{\ind}{\stackrel{ind}{\sim}}
\providecommand{\ov}[1]{\overline{#1}}


% title slide
\institute[ISU]{Department of Statistics\\Iowa State University}
\date{July 16, 2019}

\title[Gaussian process emulator]{Emulator for Water Erosion Prediction Project computer model using Gaussian Processes with functional inputs}
\author{Gulzina Kuttubekova}

\begin{document}

\begin{frame}
\maketitle

{\small
Major Professor: Dr. Jarad Niemi
}

\vspace{0.2in}
\end{frame}


% outline slide
\begin{frame}
\frametitle{Outline}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Introduction
\item Data
\item Methods
\item Analysis
\item Discussion
\item Q\&A
\end{itemize}
}

\end{frame}





% chapter: Introduction
\section{Introduction}

% slide 1
\begin{frame}
\frametitle{Computer models and Emulators}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Write the concepts of computer models and emulators
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Gaussian Processes with functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Concepts of GPs with functional inputs
}
\end{frame}





% chapter: Data
\section{Data}

% slide 1
\begin{frame}
\frametitle{WEPP computer model}
\setkeys{Gin}{width=0.7\textwidth}

{\small
What is WEPP?
}
\end{frame}


% slide 2
\subsection{WEPP simulation analysis}
\begin{frame}
\frametitle{WEPP simulation analysis}
\setkeys{Gin}{width=0.7\textwidth}

{\small
How to run WEPP? Outputs?
}
\end{frame}


% slide 3
\begin{frame}
\frametitle{Inputs: Climate and Management}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Discuss with pictures
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Inputs: Slope and Soil}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Discuss with pictures
}
\end{frame}





% chapter: Methods
\section{Methods}

% slide 1
\subsection{Gaussian Processes}
\begin{frame}
\frametitle{Gaussian process model}
\setkeys{Gin}{width=0.7\textwidth}

{\small
GP's explained: definition, covariance functions, bayesian, priors
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Properties of Gaussian process}
\setkeys{Gin}{width=0.7\textwidth}

{\small
regulzarization? penalty? marginalization?
}
\end{frame}


% slide 3
\begin{frame}
\frametitle{Gaussian process with functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
trigonometric basis? 
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Gaussian process emulator}
\setkeys{Gin}{width=0.7\textwidth}

{\small
write three final approaches 
}
\end{frame}


% chapter: Analysis
\section{Analysis}

% slide 1
\subsection{Simulation study}
\begin{frame}
\frametitle{Simulating data}
\setkeys{Gin}{width=0.7\textwidth}

{\small
how did you simulate data?
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Analysis on simulated data}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 3
\subsection{Analysis on the WEPP generated data}
\begin{frame}
\frametitle{Gaussian process model with scalar inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
description and pictures
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Gaussian process model with scalar and functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
description and pictures 
}
\end{frame}


% slide 5
\begin{frame}
\frametitle{Gaussian process model with functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
description and pictures
}
\end{frame}





% chapter: Discussion
\section{Discussion}

% slide 1
\subsection{Model performance}
\begin{frame}
\frametitle{Simulation study}
\setkeys{Gin}{width=0.7\textwidth}

{\small
how? did we expect?
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Gaussian process model with scalar inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 3
\begin{frame}
\frametitle{Gaussian process model with scalar and functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Gaussian process model with functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 5
\subsection{Next steps}
\begin{frame}
\frametitle{Sophisticating the GP emulator}
\setkeys{Gin}{width=0.7\textwidth}

{\small
how? did we expect?
}
\end{frame}


% slide 6
\begin{frame}
\frametitle{GP emulator for APSIM}
\setkeys{Gin}{width=0.7\textwidth}

{\small
what to include? improve? consider?
}
\end{frame}






% thanks slide
\begin{frame}
\frametitle{Acknowledgements}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Special thanks to my major professor Dr. Jarad Niemi, to the committee members: Dr. Somak Dutta and Dr. Max Morris and my colleagues at C-CHANGE research group
\item This slide, CC thesis, Data and R codes are available at \url{https://github.com/kgulzina/CC-Research}
\end{itemize}
}

\vspace{0.1in} \pause

% should I include references????
Manuscripts:
{\footnotesize
\begin{itemize}
\item Jarad Niemi, Eric Mittman, Will Landau, and Dan Nettleton. (2015) ``Empirical Bayes analysis of RNA-seq data for detection of gene expression heterosis.’’ Journal of Agricultural, Biological, and Environmental Statistcs, 20(4): 614-628
\item Will Landau and Jarad Niemi. ``A fully Bayesian strategy for high-dimensional hierarchical modeling using massively parallel computing.'' \url{https://arxiv.org/abs/1606.06659}
\item Will Landau, Jarad Niemi, and Dan Nettleton. ``Fully Bayesian analysis of RNA-seq counts for the detection of gene expression heterosis.'' \emph{Journal of the American Statistical Association} (in press)
\end{itemize}
}

\vspace{0.1in} \pause

\begin{center}
{\Huge
Any Questions?
}
\end{center}

\end{frame}








\subsection{Fully Bayes}
\begin{frame}
\frametitle{Priors}

All priors are constructed to be vague, proper, and (if possible) conditionally conjugate. \pause 
There are $2(L+1)$ hyperparameters and we assign the following priors
\[ \begin{array}{rll}
\tau &\sim Ga(a,b) & \mbox{conditionally conjugate} \\
\nu &\sim Unif(0,d) \\
\theta_\ell &\ind N(0,c_\ell^2) & \mbox{conditionally conjugate} \\
\sigma_\ell &\ind Unif(0,s_\ell) &
\end{array} \]
\pause
As we'll see, posterior distributions for these parameters are extremely tight relative to their priors. 

\end{frame}





\begin{frame}
\frametitle{Constructing a Gibbs sampler}

Conditional independence within a step:
\[ \begin{array}{rl}
p(\varepsilon|\ldots) &\propto \prod_{g=1}^G \prod_{n=1}^N Po\left(y_{gn}\left|e^{h_n+\varepsilon_{gn}+x_n'\beta_g}\right.\right)N(\varepsilon_{gn}|0,\gamma_g) \\
p(\gamma|\ldots) &\propto \prod_{g=1}^G \prod_{n=1}^N N(\varepsilon_{gn}|0,\gamma_g) IG(\gamma_g|\nu/2,\nu\tau/2) \\
p(\beta_{\ell}|\ldots) &\propto \prod_{g=1}^G \prod_{n=1}^N Po\left(y_{gn}\left|e^{h_n+\varepsilon_{gn}+x_n'\beta_g}\right.\right)N(\beta_{g\ell}|\theta_\ell, \sigma_\ell^2) 
\end{array} \]

\vspace{0.1in} \pause

Sufficient ``statistics'':
\[ \begin{array}{rl@{\qquad}rl}
p(\tau|\ldots) &\sim Ga(\tau|a',b') & (a',b') &= f_{\tau}(\gamma,\nu,a,b) \\
p(\nu|\ldots) &\sim p(\nu|d')\mathrm{I}(0<\nu<d) & d' &= f_\nu(\gamma,\tau,d) \\
p(\theta_\ell|\ldots) &\sim N(\theta_\ell|m'_\ell, C'_\ell) & (m'_\ell,C'_\ell) &= f_{\theta_\ell}(\beta_\ell,\sigma_\ell,c_\ell^2) \\
p(\sigma_\ell^2|\ldots) &\sim IG(e',f') \mathrm{I}(0<\sigma_\ell^2<s^2_\ell) & (e',f') &= f_{\sigma_\ell}(\beta_\ell,\theta_{\ell})
\end{array} \]
where all functions can be written as sums over $G$ terms in order to calculate
means, variances, etc.
\end{frame}


\subsection{GPUs}


\begin{frame}
\frametitle{Parallelization translated to a GPU}

If there are $G$ nodes, then 
\begin{itemize}
\item Conditional independence $\to$ embarrassingly parallel - possible speedup is $G$ \pause
\item Calculate sufficient ``statistics'' $\to$ parallel reduction - possible speedup is $[G-1]/\log_2(G)$
\end{itemize}

\vspace{0.1in}

\begin{center}
\includegraphics[width=0.6\textwidth]{parallel_reduction}

{\tiny (\url{https://scs.senecac.on.ca/~gpu610/pages/images/parallel_reduction.png})}
\end{center}
\end{frame}





\begin{frame}[fragile]
\frametitle{Implementation}
The computation for this hierarchical overdispersed count regression model is provided in the following three R packages at \url{https://github.com/wlandau/}:

\begin{itemize}
\item {\tt fbseq}: user interface 
\item {\tt fbseqOpenMP}: multithreaded backend
\item {\tt fbseqCUDA}: NVIDIA GPU backend
\end{itemize}

\pause

<<setup,size="tiny", cache=TRUE>>=
library(fbseq)
data(paschold) # Paschold et. al. (2012) data

paschold@contrasts[[5]]
paschold@contrasts[[6]]
paschold@propositions$`high-parent_B73xMo17`
@

<<dependson="setup",eval=FALSE, size="tiny">>=
configs    = Configs(burnin = 10, iterations = 10, thin = 1) 
chain      = Chain(paschold, configs) 
chain_list = fbseq(chain, backend = "CUDA")
@
\end{frame}






\section{Real data analysis}
\begin{frame}
\setkeys{Gin}{width=0.7\textwidth}
\frametitle{Analysis of Paschold et. al. (2012) data}

\begin{center}
\includegraphics{heterosis}
\end{center}

\begin{itemize}
\item $N=16$ with 4 replicates/variety on 2 plates
	\begin{itemize}
	\item varieties: B73, Mo17, B73$\times$Mo17, Mo17$\times$ B73
	\end{itemize}
\item $G=39\,656$
% \item 21\% of genes have mean counts less than 1
% \item 39\% have mean counts less than 10
% \item count: median 37, mean $260$, and max $38\,010$
\end{itemize}

\end{frame}


\begin{frame}
\frametitle{Hyperparameter posterior}
\setkeys{Gin}{width=0.8\textwidth}
\begin{center}
\includegraphics{fig-hyperhist}
\end{center}
\end{frame}


\begin{frame}
\setkeys{Gin}{width=0.6\textwidth}
\frametitle{Gene-specific parameter posteriors}

Compare posterior distribution (gray area) to normal-based approximation (black dashed line).

\vspace{-0.15in}

\begin{center}
\includegraphics{fig-betahist}
\end{center}
\end{frame}


\begin{frame}
\frametitle{Smokestack plot}
\setkeys{Gin}{width=0.8\textwidth}

% Effect size for HPH BM is the maximum of 0 and $\min(2\beta_{g2}+\beta_{g4}, 2\beta_{g3}+\beta_{g4})/\sqrt{\gamma_g}$. 

\begin{center}
\includegraphics{fig-volcano}
\end{center}
\end{frame}


\end{document}
