\documentclass[handout,10pt]{beamer}

\usetheme{AnnArbor}
\usecolortheme{beaver}

\graphicspath{{include/}}

\setlength{\unitlength}{\textwidth}  % measure in textwidths
\usepackage[normalem]{ulem}
\usepackage{graphicx}
\usepackage{subfig, accents, calc, amsthm, amsmath, amssymb, amsbsy}


% to add double tilde
\newcommand{\dbtilde}[1]{\accentset{\approx}{#1}}
\newcommand{\vardbtilde}[1]{\tilde{\raisebox{0pt}[0.85\height]{$\tilde{#1}$}}}


% to avoid counting all pages
\newcommand{\backupbegin}{
   \newcounter{finalframe}
   \setcounter{finalframe}{\value{framenumber}}
}
\newcommand{\backupend}{
   \setcounter{framenumber}{\value{finalframe}}
}


\setbeamertemplate{navigation symbols}{}
\setbeamertemplate{enumerate items}[default]
\setbeamertemplate{enumerate subitem}{\alph{enumii}.}
\setbeamertemplate{enumerate subsubitem}{\roman{enumiii}.}
\setkeys{Gin}{width=0.6\textwidth}

\newcommand{\ind}{\stackrel{ind}{\sim}}
\providecommand{\ov}[1]{\overline{#1}}


% title slide
\institute[ISU]{Department of Statistics\\Iowa State University}
\date{July 16, 2019}

\title[Gaussian process emulator]{Emulator for Water Erosion Prediction Project computer model using Gaussian Processes with functional inputs}
\author{Gulzina Kuttubekova}

\begin{document}

\begin{frame}
\maketitle

{\small
Major Professor: Dr. Jarad Niemi
}

\vspace{0.2in}
\end{frame}


% outline slide
\begin{frame}
\frametitle{Outline}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Introduction
\item Data
\item Methods
\item Analysis
\item Discussion
\item Q\&A
\end{itemize}
}

\end{frame}






% chapter: Introduction
\section{Introduction}

% slide 1
\begin{frame}
\frametitle{Concepts of computer models and emulators}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Computer models
\begin{itemize}
\item Implementation of complex mathematical models used to study many areas of scientific research (Sacks et al., 1989).
\item Process of running large computer codes with various inputs to learn the output of dynamic real systems.
\item They produce outputs that are functions of time as well as they require time-indexed inputs (Morris, 2014).
\end{itemize}

\vspace{0.1in} \pause

Emulators
\begin{itemize}
\item ``Meta-model" or ``surrogate" built with less number of inputs
\item Mimics the computer model and predicts outputs for the new input values
\end{itemize}
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Gaussian Processes (GPs)}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Collection of time- or space- dependent random variables, where any finite collection of such random variables gives the multivariate normal distribution
\item GPs provide an alternative approach to regression and classifcation problems. Derives distribution over all functions $f(x)$, which are consistent wih observed data
\item GPs employ Bayesian approach as we set prior distribution for functions,  update it with the observed data and get the posterior distribution over functions
\item Stationarity, smoothness and other aspects of GPs are controlled by tuning parameters in the covariance function $k(x,x')$
\item GPs by their structure are one of the suitable models for time- or space- varying functional input computer models
\end{itemize}
}
\end{frame}






% chapter: Data
\section{Data}

% slide 1
\begin{frame}
\frametitle{WEPP computer model}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Soil erosion and runoff prediction technology used by the United States Department of Agriculture (USDA)
\item Process-orinted, computer implemented, continuous simulation program
\item Requires various input files $\{climate, management, soil, slope\}$
\end{itemize}
}

\begin{center}
\includegraphics[scale = 0.7]{wepp.jpg}
\end{center}
%{\fontsize{2.5}{4}\selectfont (\url{https://scs.senecac.on.ca/~gpu610/pages/images/parallel_reduction.png})}
\end{frame}


% slide 2
\subsection{WEPP simulation analysis}
\begin{frame}
\frametitle{WEPP simulation analysis}
\setkeys{Gin}{width=0.7\textwidth}

\begin{center}
\includegraphics[scale = 0.1, width = 60mm]{strips.png}
\end{center}

{\small
Run WEPP separate for each hillslope, in total for 36 hillslopes in 12 watersheds at Neal Smith National Wildlife Refuge (NSNWR)
}

\end{frame}


% slide 3
\begin{frame}
\frametitle{Input files: Climate and Management}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Both files are constant for all 36 WEPP simulations
\item Climate file is retrieved from neareast station close to NSNWR
\item Precipitation is chosen amount among other daily climate records 
\begin{center}
\includegraphics{prcp.pdf}
\end{center}
\item ``corn-soybean-no-till'' management file along with the WEPP's default parameter values
\end{itemize}
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Input files: Slope and Soil}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Both files are individual for each specific hillslope
\item Soil information is retrieved from SSURGO database by USDA-NRCS
\item Soil texture is allocated in 12 watersheds by \textit{(long, lat)} coordinates provided in data files
\begin{table}[h!]
  \begin{center}
    \label{tab:table1}
    \begin{tabular}{||c c c c c c||} 
    \hline
    Depth(mm) & Sand(\%) & Clay(\%) & Organic(\%) & CEC(meq/100g) & Rock(\%) \\ [0.5ex] 
    \hline\hline
    254.00 & 6.40 & 24.50 & 2.50 & 24.00 & 0.00\\ 
    \hline
    1143.00 & 22.00 & 39.00 & 0.83 & 31.20 & 0.00\\
    \hline
    1524.00 & 25.70 & 28.00 & 0.28 & 22.40 & 0.00\\ [1ex] 
    \hline
    \end{tabular}
    \caption{Detailed LADOGA(SIL) soil texture information measured to a maximum depth of 1.8 meters.}
  \end{center}
\end{table}
\end{itemize}
}
\end{frame}


% slide 5
\begin{frame}
\frametitle{Input files: Slope and Soil}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Slope is measured for every 5 meters of hillslope length. Lengths range from 5.98 to 125.60 meters
\item ``Face''/``gulley'': Slope profile is depicted from ``face'' transect
\begin{figure}[!tbp]
  \centering
  \subfloat[Slope profile of the hillslope 1 in Basswood1 watershed.]{\includegraphics[width=0.4\textwidth]{bsw_slope.pdf}\label{fig:f1}}
  \hfill
  \subfloat[Slope profile of hillslope 3 in Orbweaver2 watershed.]{\includegraphics[width=0.4\textwidth]{orb_slope.pdf}\label{fig:f2}}
\end{figure}
\end{itemize}
}
\end{frame}


% slide 6
\begin{frame}
\frametitle{Reading information on soil loss from *.env output file}
\setkeys{Gin}{width=0.7\textwidth}

\begin{figure}[!tbp]
  \centering
  \subfloat[Anuual soil loss distribution for two random hillslopes]{\includegraphics{soil_loss.pdf}\label{fig:f1}}
\end{figure}

\end{frame}


% slide 7
\begin{frame}
\frametitle{Reading information on soil loss from *.env output file}
\setkeys{Gin}{width=0.7\textwidth}
\begin{figure}[!tbp]
  \centering
  \subfloat[Annual soil loss distribution for all 36 hillslopes in 12 years]{\includegraphics{soil_loss_all.pdf}\label{fig:f2}}
\end{figure}

\end{frame}






% chapter: Methods
\section{Methods}

% slide 1
\subsection{Gaussian Processes}
\begin{frame}
\frametitle{Gaussian process model}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item \textit{Definition}: Collection of random variables, any finite number of which have a Gaussian distribution
$$f \sim GP_x(\mu, k)$$ denote $f$ is a GP with input space $x$ s.t. for some $i,j \leq n$ we have
\begin{itemize}
\item for some mean function $\mu(x)$: $$E[f(x_i)] = \mu(x_i)$$
\item for some covariance function $k(x,x')$: $$Cov(f(x_i), f(x_j)) = k(x_i, x_j)$$
\item Using k(x,x') we build a covariance matrix $\Sigma$ s.t. for some $i,j \leq n$ $$\Sigma_{i,j} = k(x_i, x_j)$$
\end{itemize}
\item Thus, for any collection $y = \{y_1, y_2, ..., y_n\}$ s.t. $f(x_i) = y_i$, by definition, we have $$y \sim N_n(\mu, \Sigma)$$
\end{itemize}
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Properties of Gaussian process}
\setkeys{Gin}{width=0.7\textwidth}

{\small
For the inference about new output value $\tilde{y}$ given the new input value $\tilde{x}$, we have
\begin{itemize}
\item Bayesian approach: derive posterior joint distribution of $\tilde{y}$ and $y$ given prior distribution $y \sim N_n(\mu, \Sigma)$ and observed data: $$\begin{pmatrix} y \\ \tilde{y} \end{pmatrix} \sim N\begin{pmatrix} 0, & \begin{pmatrix} \Sigma & \widetilde{\Sigma} \\ \widetilde{\Sigma}^T & \dbtilde{\Sigma} \end{pmatrix} \end{pmatrix}$$
\item Predicition: conditional distribution of $\tilde{y}$ on the prior and observed data:  $$\tilde{y}|\tilde{x},x,y \sim N(\widetilde{\Sigma}^T\Sigma^{-1}y, \dbtilde{\Sigma} - \widetilde{\Sigma}^T\Sigma^{-1}\widetilde{\Sigma})$$
\item Empirical Bayes: get the MLE estimates of hyperparameters in covariance function using observed data
\end{itemize}
}
\end{frame}


% slide 3
\subsection{Gaussian processes with functional inputs}
\begin{frame}
\frametitle{Weight parameters built on trigonometric basis functions}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item \textit{Correlation distance}: for some $i,j \leq n$ and $x_i, x_j \in \mathbb{R}^T$ we have $$d(x_i, x_j) = \omega(x_i-x_j)^\alpha$$
\item \textit{Weight parameter}: $\omega \mathbb{R}^T$ s.t. $\omega_t \in \mathbb{R}^+$
\begin{itemize}
\item From WEPP, $\omega$ has periodic, wiggly structure
\item Remedy for the "curse of dimensionality" when $T >> n$: $$\omega = \alpha_0 e_0 + \sum_{q=1}^{T/2-1}(\alpha_qc_q+\beta_qs_q) + \alpha_{T/2}c_{T/2}$$ $$e_0 = (1,1, ... ,1)'$$ $$...$$ $$c_q = (\cos w_q, \cos 2w_q, ... , \cos Tw_q)'$$ $$s_q = (\sin w_q, \sin 2w_q, ... , \sin Tw_q)'$$ where $q$ is the number of harmonics
\end{itemize}
\end{itemize}
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Gaussian process emulator}
\setkeys{Gin}{width=0.7\textwidth}

{\small
By setting $\alpha = 2$ i.e \textit{Squared-exponential} covariance function, $\mu(x) = 0$ and $q = 1$ we have the following generalized model: 
$$y = \{y_1, y_2, ..., y_n\} \sim N_n(0, \Sigma)$$ for $x_i \in \mathbb{R}^D$, where $x_i = (x_{i,1}, x_{i,2}, ..., x_{i,D^{\ast}})$ and $x_{i,d} \in \mathbb{R}^{T_d}$ s.t. $1 \leq T_d \leq D$ and $\sum_{d = 1}^{D^\ast}T_d = D$
\begin{itemize}
\item Corresponding $\omega = (\omega_1, \omega_2, ..., \omega_D^\ast)$ has the same structure with $\omega_d \in \mathbb{R}^{T_d}$
\item If $T_d = 1$ input $x_{i,d}$ is scalar and functional for $T_d > 1$
\end{itemize}

\vspace{0.1in} \pause

\textit{Ridge penalty}: by adding \textit{nugget} parameter - $\lambda$ to the principal diagonal of $\Sigma$, we regularize the covariance matrix
$$\theta = (\omega_1, \omega_2, ..., \omega_D^\ast, \lambda)$$ transformed into $$\theta = (\alpha_{1,0}, \alpha_{1,1}, \beta_{1,1}, ..., \alpha_{D^\ast,0}, \alpha_{D^\ast,1}, \beta_{D^\ast,1}, \lambda)$$
}
\end{frame}






% chapter: Analysis
\section{Analysis}

% slide 1
\subsection{Simulation study}
\begin{frame}
\frametitle{Simulating data}
\setkeys{Gin}{width=0.7\textwidth}

{\small

}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Analysis on simulated data}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 3
\subsection{Analysis on the WEPP generated data}
\begin{frame}
\frametitle{Gaussian process model with scalar inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
description and pictures
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Gaussian process model with scalar and functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
description and pictures 
}
\end{frame}


% slide 5
\begin{frame}
\frametitle{Gaussian process model with functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
description and pictures
}
\end{frame}





% chapter: Discussion
\section{Discussion}

% slide 1
\subsection{Model performance}
\begin{frame}
\frametitle{Simulation study}
\setkeys{Gin}{width=0.7\textwidth}

{\small
how? did we expect?
}
\end{frame}


% slide 2
\begin{frame}
\frametitle{Gaussian process model with scalar inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 3
\begin{frame}
\frametitle{Gaussian process model with scalar and functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 4
\begin{frame}
\frametitle{Gaussian process model with functional inputs}
\setkeys{Gin}{width=0.7\textwidth}

{\small
Analysis? which model?
}
\end{frame}


% slide 5
\subsection{Next steps}
\begin{frame}
\frametitle{Sophisticating the GP emulator}
\setkeys{Gin}{width=0.7\textwidth}

{\small
how? did we expect?
}
\end{frame}


% slide 6
\begin{frame}
\frametitle{GP emulator for APSIM}
\setkeys{Gin}{width=0.7\textwidth}

\begin{center}
\includegraphics[width = 73mm, height = 71mm]{apsim}
\end{center}
% include citation
{\fontsize{2.5}{4}\selectfont (\url{https://ars.els-cdn.com/content/image/1-s2.0-S0378429018321658-gr9_lrg.jpg})}

\end{frame}






% thanks slide, Q&A
\begin{frame}
\frametitle{Acknowledgements}
\setkeys{Gin}{width=0.7\textwidth}

{\small
\begin{itemize}
\item Special thanks to my major professor Dr. Jarad Niemi, to the committee members: Dr. Somak Dutta and Dr. Max Morris and my colleagues at C-CHANGE research group
\item This slide, CC thesis, Data and R codes are available at \url{https://github.com/kgulzina/CC-Research}
\end{itemize}
}

\vspace{0.1in} \pause

\begin{center}
{\Huge
Any Questions?
}
\end{center}

\end{frame}





% references
\begin{frame}
\frametitle{References}
\setkeys{Gin}{width=0.7\textwidth}
{\footnotesize
\begin{itemize}
\item Jarad Niemi, Eric Mittman, Will Landau, and Dan Nettleton. (2015) ``Empirical Bayes analysis of RNA-seq data for detection of gene expression heterosis.’’ Journal of Agricultural, Biological, and Environmental Statistcs, 20(4): 614-628
\item Will Landau and Jarad Niemi. ``A fully Bayesian strategy for high-dimensional hierarchical modeling using massively parallel computing.'' \url{https://arxiv.org/abs/1606.06659}
\item Will Landau, Jarad Niemi, and Dan Nettleton. ``Fully Bayesian analysis of RNA-seq counts for the detection of gene expression heterosis.'' \emph{Journal of the American Statistical Association} (in press)
\end{itemize}
}
\end{frame}





\begin{frame}
\frametitle{Constructing a Gibbs sampler}

Conditional independence within a step:
\[ \begin{array}{rl}
p(\varepsilon|\ldots) &\propto \prod_{g=1}^G \prod_{n=1}^N Po\left(y_{gn}\left|e^{h_n+\varepsilon_{gn}+x_n'\beta_g}\right.\right)N(\varepsilon_{gn}|0,\gamma_g) \\
p(\gamma|\ldots) &\propto \prod_{g=1}^G \prod_{n=1}^N N(\varepsilon_{gn}|0,\gamma_g) IG(\gamma_g|\nu/2,\nu\tau/2) \\
p(\beta_{\ell}|\ldots) &\propto \prod_{g=1}^G \prod_{n=1}^N Po\left(y_{gn}\left|e^{h_n+\varepsilon_{gn}+x_n'\beta_g}\right.\right)N(\beta_{g\ell}|\theta_\ell, \sigma_\ell^2) 
\end{array} \]

\vspace{0.1in} \pause

Sufficient ``statistics'':
\[ \begin{array}{rl@{\qquad}rl}
p(\tau|\ldots) &\sim Ga(\tau|a',b') & (a',b') &= f_{\tau}(\gamma,\nu,a,b) \\
p(\nu|\ldots) &\sim p(\nu|d')\mathrm{I}(0<\nu<d) & d' &= f_\nu(\gamma,\tau,d) \\
p(\theta_\ell|\ldots) &\sim N(\theta_\ell|m'_\ell, C'_\ell) & (m'_\ell,C'_\ell) &= f_{\theta_\ell}(\beta_\ell,\sigma_\ell,c_\ell^2) \\
p(\sigma_\ell^2|\ldots) &\sim IG(e',f') \mathrm{I}(0<\sigma_\ell^2<s^2_\ell) & (e',f') &= f_{\sigma_\ell}(\beta_\ell,\theta_{\ell})
\end{array} \]
where all functions can be written as sums over $G$ terms in order to calculate
means, variances, etc.
\end{frame}


\subsection{GPUs}


\begin{frame}
\frametitle{Parallelization translated to a GPU}

If there are $G$ nodes, then 
\begin{itemize}
\item Conditional independence $\to$ embarrassingly parallel - possible speedup is $G$ \pause
\item Calculate sufficient ``statistics'' $\to$ parallel reduction - possible speedup is $[G-1]/\log_2(G)$
\end{itemize}

\vspace{0.1in}

\begin{center}
\includegraphics[width=0.6\textwidth]{parallel_reduction}

{\tiny (\url{https://scs.senecac.on.ca/~gpu610/pages/images/parallel_reduction.png})}
\end{center}
\end{frame}





\begin{frame}[fragile]
\frametitle{Implementation}
The computation for this hierarchical overdispersed count regression model is provided in the following three R packages at \url{https://github.com/wlandau/}:

\begin{itemize}
\item {\tt fbseq}: user interface 
\item {\tt fbseqOpenMP}: multithreaded backend
\item {\tt fbseqCUDA}: NVIDIA GPU backend
\end{itemize}

\pause

<<setup,size="tiny", cache=TRUE>>=
library(fbseq)
data(paschold) # Paschold et. al. (2012) data

paschold@contrasts[[5]]
paschold@contrasts[[6]]
paschold@propositions$`high-parent_B73xMo17`
@

<<dependson="setup",eval=FALSE, size="tiny">>=
configs    = Configs(burnin = 10, iterations = 10, thin = 1) 
chain      = Chain(paschold, configs) 
chain_list = fbseq(chain, backend = "CUDA")
@
\end{frame}


\end{document}
